### PR TITLE
EOS-14652 SW Update with merge algorithm

### PR DESCRIPTION
**Pull request by shalakadharap**

### CODE REVIEW

**Comment by shalakadharap** [comment]
http://eos-jenkins.colo.seagate.com/job/Cortx-PR-Build/job/S3server-7.8.2003/187/

File: rpms/s3/s3rpm.spec [hunk 1]
```diff
@@ -336,14 +355,7 @@ rm -rf %{buildroot}
...
 %post
-[ -f /opt/seagate/cortx/s3/conf/s3config.yaml ] || 
-    cp /opt/seagate/cortx/s3/conf/s3config.yaml.sample /opt/seagate/cortx/s3/conf/s3config.yaml
-[ -f /opt/seagate/cortx/s3/s3backgrounddelete/config.yaml ] ||
-    cp /opt/seagate/cortx/s3/s3backgrounddelete/config.yaml.sample /opt/seagate/cortx/s3/s3backgrounddelete/config.yaml
-[ -f /opt/seagate/cortx/auth/resources/authserver.properties ] ||
-    cp /opt/seagate/cortx/auth/resources/authserver.properties.sample /opt/seagate/cortx/auth/resources/authserver.properties
-[ -f /opt/seagate/cortx/auth/resources/keystore.properties ] ||
-    cp /opt/seagate/cortx/auth/resources/keystore.properties.sample /opt/seagate/cortx/auth/resources/keystore.properties
+sh /opt/seagate/cortx/auth/scripts/swupdate/merge.sh
```
Comment (bkirunge7) [inline rpms/s3/s3rpm.spec:358]: we dont check return value of sh /opt/seagate/cortx/auth/scripts/swupdate/merge.sh, what if it fails to merge, in that case its should go back to old version

File: scripts/swupdate/merge.sh [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+then
+    sed -i "s+$key=$configValue+$key=$newSampleValue+" $configFile
+else
+    sed -i "s+$key: $configValue+$key: $newSampleValue+" $configFile
+fi
+fi
+fi
+fi
+done < $newSampleFile
+echo "Config merge complete"
```
Comment (bkirunge7) [inline scripts/swupdate/merge.sh:52]: print file name

File: rpms/s3/s3rpm.spec (hunk 1 above)
Comment (shalakadharap) [reply rpms/s3/s3rpm.spec:358]: Rollback cas not yet handled.

File: scripts/swupdate/merge.sh (hunk 2 above)
Comment (shalakadharap) [reply scripts/swupdate/merge.sh]: Done

**Review by bkirunge7 — APPROVED** [review]

File: scripts/swupdate/merge.sh [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+#if config and old sample value same or new parameter added then replace config value with new sample value
+#else keep config value as it is
+if [ "$configValue" = "$oldSampleValue" ] || [ -z "$configValue" ]
+then
+IFS=$'\n'
+if [ "$fileType" = "properties" ]
+then
+    sed -i "s+$key=$configValue+$key=$newSampleValue+" $configFile
+else
+    sed -i "s+$key: $configValue+$key: $newSampleValue+" $configFile
```
Comment (t7ko-seagate) [inline scripts/swupdate/merge.sh:46]: This format only mentions 1 space between "key: value".  If ever anyone puts two or more spaces, this replace will silently fail.  I suggest to modify this to support arbitrary number of spaces.

File: scripts/swupdate/merge.sh [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+oldSampleValue=$(getProperty "$key" "$oldSampleFile" "$fileType")
+newSampleValue=$(getProperty "$key" "$newSampleFile" "$fileType")
+#if config and old sample value same or new parameter added then replace config value with new sample value
+#else keep config value as it is
+if [ "$configValue" = "$oldSampleValue" ] || [ -z "$configValue" ]
+then
+IFS=$'\n'
+if [ "$fileType" = "properties" ]
+then
+    sed -i "s+$key=$configValue+$key=$newSampleValue+" $configFile
```
Comment (t7ko-seagate) [inline scripts/swupdate/merge.sh:44]: I am not sure if properties file supports spaces around "=" sign.  If it does, need to modify this command to support spaces.  E.g. if config file has this entry "myKey = myVal", with spaces around "=", current implementation is not able to merge such line.

File: scripts/swupdate/merge.sh [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+read -ra ID <<< "$line"
+key="${ID[0]// /}" #remove leading and trailing spaces
+value="${ID[1]// /}"
+if ! grep -Fx "${key}" $unsafeAttributesFile >/dev/null
+then
+if [[ $key != \#* ]]; then #Avoiding commented lines or keys
+configValue=$(getProperty "$key" "$configFile" "$fileType")
+oldSampleValue=$(getProperty "$key" "$oldSampleFile" "$fileType")
+newSampleValue=$(getProperty "$key" "$newSampleFile" "$fileType")
+#if config and old sample value same or new parameter added then replace config value with new sample value
```
Comment (t7ko-seagate) [inline scripts/swupdate/merge.sh:37]: In the code below, not sure how it handles the "new added parameter".  If there is new parameter in newSampleFile, we have to add new line in configFile -- but it is not being done.

**Review by t7ko-seagate — CHANGES_REQUESTED** [review]
Pls check comments.

File: scripts/swupdate/merge.sh (hunk 5 above)
Comment (shalakadharap) [reply scripts/swupdate/merge.sh:37]: Handled. please check.

File: scripts/swupdate/merge.sh (hunk 4 above)
Comment (shalakadharap) [reply scripts/swupdate/merge.sh:49]: Yes but I am not sure how to handle this. Need your help only for this.

File: scripts/swupdate/merge.sh (hunk 3 above)
Comment (shalakadharap) [reply scripts/swupdate/merge.sh:51]: Yes but I am not sure how to handle this. Need your help only for this.

Comment (t7ko-seagate) [reply scripts/swupdate/merge.sh:51]: This should work, checked locally:

```
sed -i "s+$key: *$configValue+$key: $newSampleValue+" $configFile
```

(Added asterics `*` after space -- `key: *`.)

File: scripts/swupdate/merge.sh (hunk 4 above)
Comment (t7ko-seagate) [reply scripts/swupdate/merge.sh:49]: This should work -- tested locally:

```
sed -i "s+$key *= *$configValue+$key=$newSampleValue+" $configFile
```

(Added space-star after equal sign: `=` --> ` *= *`.

File: scripts/swupdate/merge.sh (hunk 5 above)
Comment (t7ko-seagate) [reply scripts/swupdate/merge.sh:37]: Thanks.

File: scripts/swupdate/merge.sh (hunk 3 above)
Comment (shalakadharap) [reply scripts/swupdate/merge.sh]: Yes its working @t7ko-seagate Have made smal change for same in getProperty-grep command too.
pls check 1s

File: scripts/swupdate/merge.sh (hunk 4 above)
Comment (shalakadharap) [reply scripts/swupdate/merge.sh]: Yes its working @t7ko-seagate Have made smal change for same in getProperty-grep command too.
pls check 1s

File: scripts/swupdate/merge.sh (hunk 3 above)
Comment (t7ko-seagate) [reply scripts/swupdate/merge.sh]: thanks, approved.

File: scripts/swupdate/merge.sh (hunk 4 above)
Comment (t7ko-seagate) [reply scripts/swupdate/merge.sh]: thanks, approved.

**Review by t7ko-seagate — APPROVED** [review]

File: scripts/swupdate/merge.sh [hunk 6]
```diff
@@ -0,0 +1,112 @@
+#!/bin/sh
```
Comment (nileshgovande) [inline scripts/swupdate/merge.sh:1]: indent

File: rpms/s3/s3rpm.spec [hunk 7]
```diff
@@ -210,6 +224,7 @@ rm -rf %{buildroot}
 /opt/seagate/cortx/auth/AuthServer-1.0-0.jar
 /opt/seagate/cortx/auth/AuthPassEncryptCLI-1.0-0.jar
 /opt/seagate/cortx/auth/startauth.sh
+/opt/seagate/cortx/auth/scripts/swupdate/merge.sh
```
Comment (nileshgovande) [inline rpms/s3/s3rpm.spec:227]: Write a comment in %files% section about config behavior of yum

**Review by nileshgovande — APPROVED** [review]

File: rpms/s3/s3rpm.spec (hunk 7 above)
Comment (shalakadharap) [reply rpms/s3/s3rpm.spec:227]: done

File: scripts/swupdate/merge.sh (hunk 6 above)
Comment (shalakadharap) [reply scripts/swupdate/merge.sh:1]: done

**Pull request merged**